content/advent-2019/pkgapi.md

```diff
@@ -0,0 +1,1922 @@
+
+#### A Library ####
+
+A library a collection of resources - usually source code - that is shared. Source code sharing can come in many forms. The most common way of doing this in Go is through packages, which the language supports [by specification](https://golang.org/ref/spec#Packages).
+
+There are other ways of sharing code as well. What follows is a contrived example to illustrate my point.
+
+Let's say I have a file (let's call it `lib.go`) in a directory called `common`.
+
+```
+func MaxInt(a, b int) int {
+     if a > b {
+     	return a
+     }
+     return b
+}
+```
+
+I start a new Go package (called `foo`), and I place it in a directory called `github.com/myusername/bar`. I copy `lib.go` from  `common` to `bar`, and rename the file in `bar` as `lib_bar.go`. Now I edit `lib_bar.go` and prepend the declaration `package foo` at the top so that the complete `lib_bar.go` is as follows:
+
+```
+package foo
+
+func MaxInt(a, b int) int {
+     if a > b {
+     	return a
+     }
+     return b
+}
+```
+
+Let's say I now start another new Go package (called 'baz'), and I place it in a directory called `bitbucket.org/myusername/quux`. I copy `lib.go` from `common` to `quux`, and rename the file in `quux` as `lib_baz.go`. I prepende the declaration `package baz` at the top so the complete `lib_baz.go` reads as follows:
+
+```
+package baz
+
+func MaxInt(a, b int) int {
+     if a > b {
+     	return a
+     }
+     return b
+}
+```
+
+Now, to take stock:
+
+* in `common` I have a malformed `.go` file called `lib.go`
+* in `github.com/myusername/bar` I have a file called `lib_foo.go`. The repository holds the package `foo`
+* in `github.com/myusername/quux` I have a file called `lib_baz.go`. The repository holds the package `baz`.
+
+Observe that I have shared the source code from `common/lib.go` into two different packages, `foo` and `baz`. Yes, the source code is shared by copying, instead of having a single source of truth, but the source code, at this point in time, is shared.
+
+It is in this sense that I use the word "library" - a library is source code that is shared.
+
+#### A Package ####
+
+In general, libraries of source code in Go are arranged in packages and modules. A package is a collection of `.go` files. Usually a package does one thing. A package may depend on another package.
+
+The astute reader will note that having `lib.go` in the above example will cause any Go project to have a compilation failure. All `.go` files must declare at the very top, what package it is used for. The declaration `package foo` is a conversation to the compiler, telling the compiler to include the file in a package.
+
+#### A Module ####
+
+If a package is a collection of files containing source code, a module is a collection of packages. Go modules were designed to solve package dependency issues. Modules in Go are defined by a `go.mod` file, which lists all the packages the module depends on.
```

Comment:
@chewxy not sure how relavent this is. There is a neat trick of having multiple `go.mod` files in a repository. The top level `go.mod` is for the main file and the deeper ones are for the is for the library(ies) or development time code.
In the top level relative replacements are needed to point to the libraries.

In the below example there are two libraries `convert` and `mark`. Convert in used in prod, mark is alone used during development (in is possible that mark could be used in a different prod setup to convert).

Convert depends on very little. Mark depends or convert and a pdf writing library. Main depends on mark, convert and opts.

I think this is a neat side effect of modules.
Hope you found this useful.

Example
I have a pdf conversion library. 
```
pdf$ tree -P "go.mod|*.go|*.proto" --prune -h
.
├── [ 517]  go.mod
├── [5.1K]  main.go
├── [4.0K]  pdfconv
│   ├── [3.5K]  bbox_capture.go
│   ├── [4.0K]  bbox_gen.go
│   ├── [3.6K]  bbox.go
│   ├── [  98]  go.mod
│   ├── [2.6K]  pagedsl.go
│   ├── [1.7K]  pdfconv.go
│   ├── [1.0K]  pdfdsl_bnf.go
│   └── [4.0K]  proto
│       ├── [ 27K]  document-def.pb.go
│       └── [1.1K]  document-def.proto
└── [4.0K]  pdfmarkup
    ├── [ 242]  go.mod
    └── [7.5K]  mark.go
```
The top go.mod
```
module bitbucket.org/gotoblink/gb-pdf

go 1.12

replace (
        bitbucket.org/gotoblink/gb-pdf/pdfconv => ./pdfconv
        bitbucket.org/gotoblink/gb-pdf/pdfmarkup => ./pdfmarkup
)

replace github.com/jpillora/opts v1.1.0 => github.com/millergarym/opts v1.1.3

require (
        bitbucket.org/gotoblink/gb-pdf/pdfconv v0.0.0-00010101000000-000000000000
        bitbucket.org/gotoblink/gb-pdf/pdfmarkup v0.0.0-00010101000000-000000000000
        github.com/golang/protobuf v1.3.2
        github.com/jpillora/opts v1.1.0
        github.com/signintech/pdft v0.2.0
)
```

pdfconvert
```
module bitbucket.org/gotoblink/gb-pdf/pdfconv

go 1.12

require github.com/golang/protobuf v1.3.2
```

pdfmark
```
module bitbucket.org/gotoblink/gb-pdf/pdfmarkup

replace bitbucket.org/gotoblink/gb-pdf/pdfconv => ../pdfconv

go 1.12

require (
        bitbucket.org/gotoblink/gb-pdf/pdfconv v0.0.0-00010101000000-000000000000
        github.com/signintech/pdft v0.2.0
)
```

Reply:
clever. Maybe on an article on modules